service/storage_service.cc

```diff
@@ -392,7 +387,12 @@ void storage_service::prepare_to_join(std::vector<inet_address> loaded_endpoints
         ss._local_host_id = local_host_id;
     }).get();
     auto features = get_config_supported_features();
-    _token_metadata.update_host_id(local_host_id, get_broadcast_address());
+    if (!replacing_a_node_with_diff_ip) {
+        // Replacing node with a different ip should own the host_id only after
+        // the replacing node becomes NORMAL status. It is updated in
+        // handle_state_normal().
+        _token_metadata.update_host_id(local_host_id, get_broadcast_address());
```

Comment:
Host id is the identify of a node. With replacing operation, the node identified by the host id is replaced by a new node. From the rest of the cluster's view, the node has no different before or after the replace operation. 

If the replacing node uses a different host id, the rest of the cluster can not know who the replacing node is replacing with. Look at storage_service::handle_state_normal how we use the host id to figure out the node becomes NORMAL is replacing another node.

Reply:
But, it is a different node in reality. The fact that we try to reuse the same host id on a different node generates special cases all around. Like in this hunk.

So I wonder, why not use a different host id like in bootstrap. The end result would be the same as if you decomissioned and then added a new node with the same tokens.

The rest of the cluster can know that we're replacing from the fact that tokens collide.

Reply:
This colliding can happen if we intend to bootstrap a node but unfortunately it generates tokens which belongs to an existing node. The other nodes would think it is replacing  while it is not. I think it is better to be explicit. 

Also, replacing node uses the same host id is something existing Cassandra user would expect.

Reply:
And, when the existing node sees the replacing node advertises to be replacing status, the existing nodes needs to look at the host id of the replacing node to know which node this replacing node is replacing with.

Reply:
Long term when we move to multi-node topology change operations I think replace should not exist, and just be a special case of remove + add. For instance if you have the cluster A, B, C, you can commit to add D E and F and then {A, B, C} -> {A, B, C, D, F}

If C is dead this is just a special case where you do {A, B, C} -> {A, B, D, F}. So at least in theory I am in favor of what Tomek suggests that maybe we should just handle this case now: let the node enters with a different ID, and we can just figure out a way to accept the fact that it will reuse the same tokens. If you think about it: bootstrap nodes can take writes, so we'll get that for free.

Can we maybe move token selection to a coordinator node instead of allowing the new node to propose its own tokens?  That's how multi-node would do anyway. If the coordinator says it is fine that the token collide, then it is fine that the tokens collide.

Reply:
> This colliding can happen if we intend to bootstrap a node but unfortunately it generates tokens which belongs to an existing node. The other nodes would think it is replacing while it is not. I think it is better to be explicit.

The bootstrapping node should not chose such tokens in the first place.

If we still want to catch this, we can look at the status of the new node and allow only "hibernate" to replace.


> And, when the existing node sees the replacing node advertises to be replacing status, the existing nodes needs to look at the host id of the replacing node to know which node this replacing node is replacing with.

This can be derived from token metadata, you replace the node which owns any token you replace in the old ring.


> 
> Also, replacing node uses the same host id is something existing Cassandra user would expect.

Do you know of any good reason a user may expect that?

I'm interested in this topic not only for this PR. It will make generic topology changes easier.

I think we should switch to identifying nodes by host id so that we don't have to special case the case of replacing a node with the same ip.

Having two nodes in the system with the same host id precludes that, because host id is still not unique (which is ironic).

If users wants the same id after replace, it could be an extra step after whole operation where you rename a node, but after you remove the old id from the system altogether. So host id is still unique. It'd be better to avoid this, still.

Reply:
The direction we are moving towards, with multi-add, is something that departs from Cassandra abruptly anyway.

What Cassandra users expect in this case is not important. We just have to document it.

Reply:
When looking at user reported cluster issue, I usually check nodetool status to see if the host_id is owned by the ip address of the new replacing node. If that is the case, I know it is a  replace operation. If the host id changes, it is a bootstrap operation.  
Sometimes, due to operation error, we might find the replacing node uses the same host id but with new and different tokens. If the replacing node would use different host_id, we do not even know it is supposed to be an replace operation but ended up using wrong tokens.


The mapping between host id and tokens never changes. It is a very good property in my view. The host id is added and removed from the cluster when new nodes are added or removed from the cluster. The whole code base assumes this. Do you have a very good reason to change this?

@tgrabiec Can you please point out what are the special cases you are talking about  "The fact that we try to reuse the same host id on a different node generates special cases all around" beyond  this
```
+    if (!replacing_a_node_with_diff_ip) {
+        // Replacing node with a different ip should own the host_id only after
+        // the replacing node becomes NORMAL status. It is updated in
+        // handle_state_normal().
+        _token_metadata.update_host_id(local_host_id, get_broadcast_address());
+    }
```

This special case is needed to avoid more special cases to happen (handling replacing of the dead node differently on the replacing node than the rest of the cluster)

Look how unified  it is with this series: 

```
1) node6 (replacing node)
INFO  2020-04-24 09:36:34,331 [shard 0] storage_service - Set host_id=159d50be-2e2f-41c5-83ec-06d0e02681f8 to be owned by node=127.0.0.2
INFO  2020-04-24 09:36:34,331 [shard 0] storage_service - Set host_id=a855d88a-41ca-47ac-857e-a0ceb41adc6c to be owned by node=127.0.0.3
INFO  2020-04-24 09:36:34,331 [shard 0] storage_service - Set host_id=19784af8-6991-445c-a1f2-fa1e6568c73d to be owned by node=127.0.0.4
INFO  2020-04-24 09:36:34,332 [shard 0] storage_service - Set host_id=faf19e83-d97c-4267-941a-9a9fc1d3cbd2 to be owned by node=127.0.0.1
INFO  2020-04-24 09:36:34,332 [shard 0] storage_service - Set host_id=80135248-ec23-4c48-8701-47cdda0abe99 to be owned by node=127.0.0.5
 

 
WARN  2020-04-24 09:38:32,142 [shard 0] storage_service - Host ID collision for 80135248-ec23-4c48-8701-47cdda0abe99 between 127.0.0.5 and 127.0.0.6; 127.0.0.6 is the new owner
INFO  2020-04-24 09:38:32,142 [shard 0] storage_service - Set host_id=80135248-ec23-4c48-8701-47cdda0abe99 to be owned by node=127.0.0.6, existing=127.0.0.5
INFO  2020-04-24 09:38:32,142 [shard 0] storage_service - Remove node 127.0.0.6 from pending replacing endpoint


2) node1 (existing node):

INFO  2020-04-24 09:36:06,191 [shard 0] storage_service - Set host_id=159d50be-2e2f-41c5-83ec-06d0e02681f8 to be owned by node=127.0.0.2
INFO  2020-04-24 09:36:08,204 [shard 0] storage_service - Set host_id=a855d88a-41ca-47ac-857e-a0ceb41adc6c to be owned by node=127.0.0.3
INFO  2020-04-24 09:36:09,329 [shard 0] storage_service - Set host_id=19784af8-6991-445c-a1f2-fa1e6568c73d to be owned by node=127.0.0.4
INFO  2020-04-24 09:36:12,342 [shard 0] storage_service - Set host_id=80135248-ec23-4c48-8701-47cdda0abe99 to be owned by node=127.0.0.5


WARN  2020-04-24 09:38:32,941 [shard 0] storage_service - Host ID collision for 80135248-ec23-4c48-8701-47cdda0abe99 between 127.0.0.5 and 127.0.0.6; 127.0.0.6 is the new owner
INFO  2020-04-24 09:38:32,941 [shard 0] storage_service - Set host_id=80135248-ec23-4c48-8701-47cdda0abe99 to be owned by node=127.0.0.6, existing=127.0.0.5
INFO  2020-04-24 09:38:32,941 [shard 0] storage_service - Remove node 127.0.0.6 from pending replacing endpoint

```

Reply:
> The direction we are moving towards, with multi-add, is something that departs from Cassandra abruptly anyway.
> 
> What Cassandra users expect in this case is not important. We just have to document it.

We can still keep the property that host_id does not change in replacing operation with multi-add. If breaking compatibility is a must we consider it, if it is not why break it.

Reply:
> > This colliding can happen if we intend to bootstrap a node but unfortunately it generates tokens which belongs to an existing node. The other nodes would think it is replacing while it is not. I think it is better to be explicit.
> 
> The bootstrapping node should not chose such tokens in the first place.
> 
> If we still want to catch this, we can look at the status of the new node and allow only "hibernate" to replace.

Bug happens and we should be defensive. Yes, the extra check would help in this case. 


> > And, when the existing node sees the replacing node advertises to be replacing status, the existing nodes needs to look at the host id of the replacing node to know which node this replacing node is replacing with.
> 
> This can be derived from token metadata, you replace the node which owns any token you replace in the old ring.

Yes, this is doable. More complicated though. We need to get the tokens of the replacing node from gossip. Then choose one of the tokens and look into token_metadata for the ip address to find out the old node. we may ask: Is it enough to look at just one token? What if a bootstrap node happen to get a same token. If simply looking at a host id is enough, why not.


> > Also, replacing node uses the same host id is something existing Cassandra user would expect.
> 
> Do you know of any good reason a user may expect that?
> 
> I'm interested in this topic not only for this PR. It will make generic topology changes easier.
> 
> I think we should switch to identifying nodes by host id so that we don't have to special case the case of replacing a node with the same ip.
> 
> Having two nodes in the system with the same host id precludes that, because host id is still not unique (which is ironic).
> 
> If the user wants the same id after replace, it could be an extra step after the whole ring operation is complete, where we rename a node, but after you remove the old id from the system altogether. So host id is still unique. It'd be better to avoid this, still.

I explained in the other comments why keeping the host_id is useful.  

Regarding to use host_id or whatever id to identify a node, we have been talking about this for a very long time. For example use some id instead of ip in gossip as well. It is far beyond what this PR solves. If you think it is important please open an github issue we can discuss there. 
I hope you do not insist it as a dependency of this PR and block this PR for 1 year and a half.

Reply:
> When looking at user reported cluster issue, I usually check nodetool status to see if the host_id is owned by the ip address of the new replacing node. If that is the case, I know it is a replace operation. If the host id changes, it is a bootstrap operation.
> Sometimes, due to operation error, we might find the replacing node uses the same host id but with new and different tokens. If the replacing node would use different host_id, we do not even know it is supposed to be an replace operation but ended up using wrong tokens.

We could satisfy this need in a different manner. Rather than reuse host id, we can make the replacing node advertise through gossip the host id which it replaces and nodetool can pick it up and present.

Even if we don't do it, we can still get this information from the log of the replacing node.

Reply:
> 
> The mapping between host id and tokens never changes. It is a very good property in my view. The host id is added and removed from the cluster when new nodes are added or removed from the cluster. The whole code base assumes this. Do you have a very good reason to change this?

I think host id is supposed to identify the host, not its tokens. If you decommission a node and add a new node with the same tokens, it will have a different host id despite having the same tokens. So a host has tokens, host id identifies a host, but host id != tokens. 

Which code do you think assumes that host id == tokens? 

If we needed a hash of host's tokens, we should introduce a new variable for that.

> 
> @tgrabiec Can you please point out what are the special cases you are talking about "The fact that we try to reuse the same host id on a different node generates special cases all around" beyond this
> 
> ```
> +    if (!replacing_a_node_with_diff_ip) {
> +        // Replacing node with a different ip should own the host_id only after
> +        // the replacing node becomes NORMAL status. It is updated in
> +        // handle_state_normal().
> +        _token_metadata.update_host_id(local_host_id, get_broadcast_address());
> +    }
> ```

It generates special cases indirectly because if we don't have a unique identifier of a node, the coordinator code which routes requests has to special case the replace operation by filtering replaced node from the pending endpoint list.

> 
> This special case is needed to avoid more special cases to happen (handling replacing of the dead node differently on the replacing node than the rest of the cluster)

Why can't we have unified handling if replacing node uses a different id?

Reply:
> > > This colliding can happen if we intend to bootstrap a node but unfortunately it generates tokens which belongs to an existing node. The other nodes would think it is replacing while it is not. I think it is better to be explicit.
> > 
> > 
> > The bootstrapping node should not chose such tokens in the first place.
> > If we still want to catch this, we can look at the status of the new node and allow only "hibernate" to replace.
> 
> Bug happens and we should be defensive. Yes, the extra check would help in this case.
> 
> > > And, when the existing node sees the replacing node advertises to be replacing status, the existing nodes needs to look at the host id of the replacing node to know which node this replacing node is replacing with.
> > 
> > 
> > This can be derived from token metadata, you replace the node which owns any token you replace in the old ring.
> 
> Yes, this is doable. More complicated though. 

Another option is to advertise the replaced host id using a separate gossip property.

> We need to get the tokens of the replacing node from gossip. Then choose one of the tokens and look into token_metadata for the ip address to find out the old node. we may ask: Is it enough to look at just one token? What if a bootstrap node happen to get a same token. If simply looking at a host id is enough, why not.

Are you thinking about concurrent bootstrap and replace here? 


> 
> > > Also, replacing node uses the same host id is something existing Cassandra user would expect.
> > 
> > 
> > Do you know of any good reason a user may expect that?
> > I'm interested in this topic not only for this PR. It will make generic topology changes easier.
> > I think we should switch to identifying nodes by host id so that we don't have to special case the case of replacing a node with the same ip.
> > Having two nodes in the system with the same host id precludes that, because host id is still not unique (which is ironic).
> > If the user wants the same id after replace, it could be an extra step after the whole ring operation is complete, where we rename a node, but after you remove the old id from the system altogether. So host id is still unique. It'd be better to avoid this, still.
> 
> I explained in the other comments why keeping the host_id is useful.
> 
> Regarding to use host_id or whatever id to identify a node, we have been talking about this for a very long time. For example use some id instead of ip in gossip as well. It is far beyond what this PR solves. If you think it is important please open an github issue we can discuss there.
> I hope you do not insist it as a dependency of this PR and block this PR for 1 year and a half.

I'm not insisting to make this a dependency yet. I'm still trying to understand if this PR is correct. This problem makes it harder to prove it correct.

Reply:
So are you planing to do so? What part of the host_id related in this series do you think it is incorrect? I am happy to hear.